activerecord/lib/active_record/type_caster/connection.rb

```diff
@@ -13,8 +14,20 @@ def type_cast_for_database(attr_name, value)
       end
 
       def type_for_attribute(attr_name)
-        @klass.type_for_attribute(attr_name) || Type.default_value
+        schema_cache = @klass.schema_cache
+
+        if schema_cache.data_source_exists?(table_name)
+          column = schema_cache.columns_hash(table_name)[attr_name.to_s]
+          if column
+            type = @klass.with_connection { |connection| connection.lookup_cast_type_from_column(column) }
```

Comment:
Note to self: this is OK for now but I need to find another solution to eliminate this `with_connection`.